### PR TITLE
feat(storage): support object retention

### PR DESCRIPTION
**Pull request by coryan**
Implement support for object retention. This introduces a new object
metadata field. Any patch or update operation that changes this field
requires an extra option (`OverrideUnlockedRetention`).

Part of the work for #12994

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14351)

### CODE REVIEW

File: google/cloud/storage/object_retention.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+namespace cloud {
+namespace storage {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Specifies the retention parameters of an object.
+ *
+ * Objects under retention cannot be deleted or overwritten until their
+ * retention expires. Objects with a "Locked" retention cannot have their
+ * retention period decreased or incce. The soft delete policy prevents
```
Comment (bajajneha27) [inline google/cloud/storage/object_retention.h:34]:
```suggestion
 * retention period decreased or increased. The soft delete policy prevents
```
Probably a typo?

File: google/cloud/storage/examples/storage_object_retention_samples.cc [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+// limitations under the License.
+
+#include "google/cloud/storage/client.h"
+#include "google/cloud/storage/examples/storage_examples_common.h"
+#include "google/cloud/internal/getenv.h"
+#include <iostream>
+
+namespace {
+
+void InsertObjectWithRetention(google::cloud::storage::Client client,
```
Comment (bajajneha27) [inline google/cloud/storage/examples/storage_object_retention_samples.cc:22]: Do we not region tags for these samples yet?

**Review by bajajneha27 — APPROVED** [review]
LGTM except for two tiny comments.

File: google/cloud/storage/object_retention.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/object_retention.h]: Fixed, thanks.

File: google/cloud/storage/examples/storage_object_retention_samples.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_object_retention_samples.cc:22]: Candidly, I did not look.  Now that I did, I only found `storage_set_object_retention_policy` and `storage_create_bucket_with_object_retention`.

**Review by coryan — COMMENTED** [review]
Thanks!

**Pull request merged**